src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs

```diff
@@ -220,22 +220,42 @@ internal void Clear()
             lock (_readSyncObject)
             {
                 priorityType = DataPriorityType.Default;
-
-                // send data from which ever stream that has data directly.
                 byte[] result = null;
-                result = _dataToBeSent[(int)DataPriorityType.PromptResponse].ReadOrRegisterCallback(_onSendCollectionDataAvailable);
-                priorityType = DataPriorityType.PromptResponse;
 
-                if (result == null)
-                {
-                    result = _dataToBeSent[(int)DataPriorityType.Default].ReadOrRegisterCallback(_onSendCollectionDataAvailable);
-                    priorityType = DataPriorityType.Default;
-                }
-                // no data to return..so register the callback.
-                if (result == null)
+                if (_dataSyncObjects != null && _dataToBeSent != null)
                 {
-                    // register callback.
-                    _onDataAvailableCallback = callback;
+                    const int promptResponseIndex = (int)DataPriorityType.PromptResponse;
+
+                    lock (_dataSyncObjects[promptResponseIndex])
+                    {
+                        //NOTE: fix unhandled NullReference exception;
+                        //_dataToBeSent[promptResponseIndex] is disposed by Clear method from different thread 
+                        if (_dataToBeSent[promptResponseIndex] != null) 
+                        {
+                            // send data from which ever stream that has data directly.
+                            result = _dataToBeSent[promptResponseIndex].ReadOrRegisterCallback(_onSendCollectionDataAvailable);
+                            priorityType = DataPriorityType.PromptResponse;
+                        }
+                    }
+
+                    if (result == null)
+                    {
+                        const int defaultIndex = (int)DataPriorityType.Default;
+                        lock (_dataSyncObjects[defaultIndex]) 
+                        {
+                            if (_dataToBeSent[defaultIndex] != null) 
```

Comment:
Please add comment here:
`// _dataToBeSent array items can be asynchronously disposed and set to null value.`